From 428e1ec1d1c451e0297032ab6f0dffe91bbfaefc Mon Sep 17 00:00:00 2001
From: z060142 <z060142@yahoo.com.tw>
Date: Sun, 30 Aug 2026 16:16:18 +0800
Subject: [PATCH 209/307] !F (3DEngine) Expose terrain integration CVars.

Use a floating-point maximum height so sub-metre bands can be configured.
Include the height and vertex limits in terrain refresh state so CVar changes
invalidate affected sector meshes.
---
 Code/CryEngine/Cry3DEngine/3dEngine.cpp |  4 +++-
 Code/CryEngine/Cry3DEngine/cvars.cpp    | 17 ++++++++++++++---
 Code/CryEngine/Cry3DEngine/cvars.h      |  3 ++-
 3 files changed, 19 insertions(+), 5 deletions(-)

diff --git a/Code/CryEngine/Cry3DEngine/3dEngine.cpp b/Code/CryEngine/Cry3DEngine/3dEngine.cpp
index 30ca3cb5..48e370b5 100644
--- a/Code/CryEngine/Cry3DEngine/3dEngine.cpp
+++ b/Code/CryEngine/Cry3DEngine/3dEngine.cpp
@@ -754,7 +754,9 @@ void C3DEngine::ProcessCVarsChange()
 		GetFloatCVar(e_ViewDistCompMaxSize) +
 		GetCVars()->e_DecalsDeferredStatic +
 		GetCVars()->e_TerrainBlendingDebug +
-		GetCVars()->e_TerrainDetailMaterialsWeightedBlending;
+		GetCVars()->e_TerrainDetailMaterialsWeightedBlending +
+		GetCVars()->e_TerrainIntegrateObjectsMaxVertices +
+		GetCVars()->e_TerrainIntegrateObjectsMaxHeight;
 
 	if (m_fRefreshSceneDataCVarsSumm != -1 && m_fRefreshSceneDataCVarsSumm != fNewCVarsSumm)
 	{
diff --git a/Code/CryEngine/Cry3DEngine/cvars.cpp b/Code/CryEngine/Cry3DEngine/cvars.cpp
index e98e8b69..504c7987 100644
--- a/Code/CryEngine/Cry3DEngine/cvars.cpp
+++ b/Code/CryEngine/Cry3DEngine/cvars.cpp
@@ -492,10 +492,21 @@ void CVars::Init()
 	REGISTER_CVAR(e_TerrainAutoGenerateBaseTextureTiling, 1.f / 16.f, VF_NULL,
 	              "Controls tiling of baked diffuse textures");
 	REGISTER_CVAR(e_TerrainIntegrateObjectsMaxVertices, 30000, VF_NULL,
-	              "Preallocate specified number of vertices to be used for objects integration into terrain (per terrain sector)\n"
-	              "0 - disable the feature completelly");
+	              "Per terrain sector vertex budget for the level flag Terrain/IntegrateObjects\n"
+	              "(objects whose GI mode is 'Integrate Into Terrain' get a copy of their LOD 0 geometry\n"
+	              "appended to the terrain sector mesh, so the terrain detail layers are painted onto them).\n"
+	              "This many vertices are preallocated for every LOD 0 sector on each sector rebuild,\n"
+	              "and objects are truncated once the budget is exhausted.\n"
+	              "0 - disable the feature completely (also clears the level flag on the next level load)");
 	REGISTER_CVAR(e_TerrainIntegrateObjectsMaxHeight, 32.f, VF_NULL,
-	              "Take only trianglses close to terrain for objects integration");
+	              "Height band used by the level flag Terrain/IntegrateObjects: only object triangles\n"
+	              "within this many meters above the terrain under them are integrated.\n"
+	              "Float; values below 1 are useful for welding flat plates onto the ground");
+	REGISTER_CVAR(e_TerrainIntegrateObjectsDebug, 0, VF_NULL,
+	              "Debug output for the level flag Terrain/IntegrateObjects\n"
+	              "0 - off\n"
+	              "1 - log the integrated vertex/triangle count of every terrain sector rebuild\n"
+	              "2 - additionally draw the bounding box of the integrated triangles of each sector");
 	DefineConstIntCVar(e_TerrainDeformations, 0, VF_CHEAT,
 	                   "Allows in-game terrain surface deformations");
 	REGISTER_CVAR(e_TerrainEditPostponeTexturesUpdate, 10, VF_NULL,
diff --git a/Code/CryEngine/Cry3DEngine/cvars.h b/Code/CryEngine/Cry3DEngine/cvars.h
index 164b1ca8..cd00d295 100644
--- a/Code/CryEngine/Cry3DEngine/cvars.h
+++ b/Code/CryEngine/Cry3DEngine/cvars.h
@@ -220,7 +220,8 @@ struct CVars : public Cry3DEngineBase
 	int   e_TerrainAutoGenerateBaseTexture;
 	float e_TerrainAutoGenerateBaseTextureTiling;
 	int   e_TerrainIntegrateObjectsMaxVertices;
-	int   e_TerrainIntegrateObjectsMaxHeight;
+	float e_TerrainIntegrateObjectsMaxHeight;
+	int   e_TerrainIntegrateObjectsDebug;
 	DeclareConstIntCVar(e_SkyBox, 1);
 	float e_CoverageBufferAABBExpand;
 	int   e_CoverageBufferEarlyOut;
-- 
2.47.1.windows.1

